### ISSUE TITLE
Proxy Collection Messages (DEF-1249)

### ISSUE THREAD

**Issue by britzl** [defold/defold#3494]
As an addition to the async loading it would be helpful if the collection gets notified on loading/unloading and enabling/disabling.

Right now, when a proxy collection is loaded nothing on the collection is called. The init() function is called the FIRST time the collection is ENABLED. Calling disable/enable again won't trigger init() again. Unloading will trigger final(). This behaviour is a bit unpredictable.

I suggest the following:
loaded -> init()
unloaded -> final()
enabled -> on_enable()
disabled -> on_disable()
2015-06-25 17:02:38 (ingela.garneij)
If you only need notifications of loading/unloading and enabling/disabling, then you can already today script these functions yourself. Or are you in need of something more that we missed?
2015-07-17 15:14:33 (Alexander.Ericsson)
Hi,

It's not really possible to do in a stable manner as it works right now. Here is the long explanation:

When calling load/async_load on a collection it will load its internals (textures, etc) and then post a proxy_loaded message to the loader (not the collection that was loaded).
At this point there is not anything that can be done with the collection. No function has been called on it (not even init) and it's not possible to post messages to it since it's not enabled.
What's more, only the internals of the collection have been loaded. Since we haven't had the opportunity to run any user code in the collection, all "our" stuff is not loaded/setup at this point.

The only option is to enable to collection. This is done by sending an (async) enable message. There is no callback and no on_enabled message once done enabling. There is one exception. The first time (after being loaded) enable is called, init() will be triggered on the collection. Any subsequent enable/disable calls will not trigger init() again. This is problematic since we want to load a collection (while displaying a loading screen), with the async_load call, then enable/disable it as many times we want without any loading-times. Then unload the collection whenever its not used anymore. At each of these four states (load/unload/enable/disable) we would like the collection to be notified (through callback or message).

Our current hacky/semi-working workarounds involve calling enable+disable next to each other right after the "on_loaded" message is received, to run our user code in init, triggered only by enable. Other things like waiting an arbitrary time (like 5 frames, 1 is too little) after enable is called before we can post messages to the collection. If "enable" for some reason uses 6 frames our message will be lost. When disabling the collection we first have to send a custom disable message so the collection run its disable code, then wait one frame, then send the real disable to disable the collection. If both messages are sent in the same frame, the collection will be disabled first and the other message gets discarded.

Having clear built-in states is key for us to build stable state switching in game.

**Comment by britzl** [comment]
Won't fix.
